backends-velox/src/main/scala/org/apache/gluten/expression/aggregate/VeloxCollect.scala

```diff
@@ -0,0 +1,65 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.gluten.expression.aggregate
+
+import org.apache.spark.sql.catalyst.expressions.{ArrayDistinct, ArrayUnion, AttributeReference, CreateArray, Expression, Literal}
+import org.apache.spark.sql.catalyst.expressions.aggregate.DeclarativeAggregate
+import org.apache.spark.sql.catalyst.trees.UnaryLike
+import org.apache.spark.sql.types.{ArrayType, DataType}
+
+abstract class VeloxCollect extends DeclarativeAggregate with UnaryLike[Expression] {
+  protected lazy val buffer: AttributeReference = AttributeReference("buffer", dataType)()
+
+  override def dataType: DataType = ArrayType(child.dataType, false)
+
+  override def aggBufferAttributes: Seq[AttributeReference] = List(buffer)
+
+  override lazy val initialValues: Seq[Expression] = List(Literal.create(Seq.empty, dataType))
+
+  override lazy val updateExpressions: Seq[Expression] = List(
+    ArrayUnion(buffer, CreateArray(List(child), useStringTypeWhenEmpty = false))
```

Comment:
For each row, we wrap a array. Will it cause performance regression with vanilla Spark ?

Reply:
Does the semaintics of `ArrayUnion` equal to `Collect` ?
- `ArrayUnion` always returns distinct value in an array
- `ArrayUnion` allows null but CollectList/CollectSet do not allow

Reply:
> For each row, we wrap a array. Will it cause performance regression with vanilla Spark ?

Yes that's possible. But I assume it should be a comparatively rare case when user enables Gluten however these two functions are fallen back. So I think it's fine until getting real complaints from users.

Reply:
> ArrayUnion always returns distinct value in an array

Oops, if it does de-dup then `velox_collect_list` may return wrong result. I'll take a look on this.

Reply:
So [concat](https://spark.apache.org/docs/latest/api/sql/index.html#concat) could be the right choice here?

Reply:
concat allows null but collect_list does not allow:
- the result is null
- the null element value in array

I think we should at least pass test `collect_set/collect_list with null` in `VeloxAggregateFunctionsSuite` with partially fallback.

Reply:
Instead of adding more cutsom Gluten expression, cann't we just add project to adapt data type (convert array to binary, etc) in columnar rule ?

Reply:
Sure, let me add more test against this case.

Reply:
@ulysses-you Please take a look again. 

Change:
https://github.com/apache/incubator-gluten/pull/5655/files#diff-1ebbabef839649a226bf6bed298759976ab8adafe31d3dbfe07c559a1a95842dR33-R38

Test case:
https://github.com/apache/incubator-gluten/pull/5655/files#diff-742ce9027f0d774d324125a130d9993bad09d4ea5eb6d35d791b39bb57138c77R983-R1057